### PR TITLE
Channel: refactor force-close handling

### PR DESCRIPTION
**Pull request by aarani**

### CODE REVIEW

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs [hunk 1]
```diff
@@ -17,3 +24,431 @@ module internal ChannelConstantHelpers =
...
+
+                let! perCommitmentSecret =
+                    match remotePerCommitmentSecret with
+                    | Choice1Of2 remotePerCommitmentSecret -> Ok remotePerCommitmentSecret
+                    | Choice2Of2 remotePerCommitmentSecretStore ->
+                        let commitmentSecretOpt =
+                            remotePerCommitmentSecretStore.GetPerCommitmentSecret commitmentNumber
+
+                        match commitmentSecretOpt with
+                        | Some commitmentSecret -> Ok commitmentSecret
```
Comment (aarani) [inline src/DotNetLightning.Core/Channel/ChannelHelpers.fs:261]: We check for revoked closure after all our other checks return false, this makes it possible that a non-remote tx (maybe an old local commitment tx) gets caught by this function, so is it possible that this causes a False positive? (maybe there's a revoked remote commitment with the same commitment number)

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs [hunk 2]
```diff
@@ -17,3 +24,431 @@ module internal ChannelConstantHelpers =
...
+
+                        transactionBuilder.Extensions.Add(CommitmentToLocalExtension())
+
+                        transactionBuilder
+                            .AddKeys(revocationPrivKey.RawKey())
+                            .AddCoin(ScriptCoin(closingTx, toLocalIndex |> uint32, toLocalScriptPubKey))
+                        |> ignore)
+
+                // FIXME: what if all the money is in htlcs?
+                if toLocalIndexOpt.IsNone && toRemoteIndexOpt.IsNone then
```
Comment (aarani) [inline src/DotNetLightning.Core/Channel/ChannelHelpers.fs:329]: This is a temporary fix for the false-positive scenario, this causes the function to return "Unknown Closing Tx" if we couldn't find any to_local and to_remote output, but this creates an issue with commitment TXs that only have HTLC outputs and not any main one.

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs [hunk 3]
```diff
@@ -17,3 +24,431 @@ module internal ChannelConstantHelpers =
...
+                    MainOutput = Error UnknownClosingTx
+                }
+
+    let HandleFundingTxSpent
+        (savedChannelState: SavedChannelState)
+        (remoteNextCommitInfoOpt: Option<RemoteNextCommitInfo>)
+        (channelPrivKeys: ChannelPrivKeys)
+        (closingTx: Transaction) : ClosingResult =
+        let closingTxId = closingTx.GetTxId()
+        if closingTxId = savedChannelState.LocalCommit.PublishableTxs.CommitTx.Value.GetTxId() then
```
Comment (aarani) [inline src/DotNetLightning.Core/Channel/ChannelHelpers.fs:430]: Is there any other way to recognize local commitments? (This current check only works for last local commitment and not previous/revoked ones)

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs (hunk 2 above)
Comment (joemphilips) [reply src/DotNetLightning.Core/Channel/ChannelHelpers.fs:329]: I want you to add test to check the false positive scenario.

>  but this creates an issue with commitment TXs that only have HTLC outputs and not any main one.

And a test to check that this won't lead to the crash?

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs (hunk 3 above)
Comment (joemphilips) [reply src/DotNetLightning.Core/Channel/ChannelHelpers.fs:430]: 🤷

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs [hunk 4]
```diff
@@ -17,3 +24,431 @@ module internal ChannelConstantHelpers =
...
+                | claimMainOutput ->
+                    {
+                        MainOutput = claimMainOutput
+                    }
+            | _ ->
+                {
+                    MainOutput = Error UnknownClosingTx
+                }
+
+    let HandleFundingTxSpent
```
Comment (joemphilips) [inline src/DotNetLightning.Core/Channel/ChannelHelpers.fs:424]: I think this function needs test?

**Comment by joemphilips** [comment]
some @aarani 's questions are hard hard for me to answer. And I have no time to catch up. (Channel state handling is not my interest anyway.)
Since at least there is no malicious code, I'd better just approve this change after tests I've mentioned above.

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs (hunk 2 above)
Comment (aarani) [reply src/DotNetLightning.Core/Channel/ChannelHelpers.fs:337]: > I want you to add test to check the false positive scenario.
I added a test for old local commitment txs 
> And a test to check that this won't lead to the crash?
I think I should explain more, the only problem with that edge case is that instead of BalanceBelowDustLimit error we return UnknownClosingTx which seems a bad error for the case (I added a comment explaining it the code also)

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs (hunk 4 above)
Comment (aarani) [reply src/DotNetLightning.Core/Channel/ChannelHelpers.fs:432]: I added a test for revoked/remote/local cases.

**Comment by knocte** [comment]
How's this looking Joe?

@canndrew hey hello, by any chance you can have a quick look in case you spot something off very quickly?

File: tests/DotNetLightning.Core.Tests/TransactionTests.fs [hunk 5]
```diff
@@ -326,4 +329,551 @@ let testList = testList "transaction tests" [
...
+            NodeMasterPrivKey extKey
+        let remoteChannelPrivKeys = remoteNodeMasterPrivKey.ChannelPrivKeys (rand.Next(1, 100))
+        let remoteChannelPubKeys = remoteChannelPrivKeys.ToChannelPubKeys()
+
+        let fundingAmount = 10_000_000L |> Money.Satoshis
+        let fundingScriptPubKey =
+            Scripts.funding
+                localChannelPubKeys.FundingPubKey
+                remoteChannelPubKeys.FundingPubKey
+        let fundingDestination = fundingScriptPubKey.WitHash :> IDestination
```
Comment (joemphilips) [inline tests/DotNetLightning.Core.Tests/TransactionTests.fs:352]: It is not super-important, but I want you to have a small scope for temporary variable like this.
i.e., instead of writing
```fsharp
let x = 1
let y = 2
let z = x + y
```

write

```fsharp
let z =
    let x = 1
    let y = 2
    x + y
```

In this way I can understand the lifetime of `x`, `y` more easily and it gets easy to review. (This is one of the biggest reason I use F#)

In this case, `fundingScriptCode` is what we want to have later. So

```fsharp

let fundingScriptCoin =
    let fundingDestination = // ..
    // ...
    ScriptCoin(fundingCoin, fundingScriptPubKey)
```

Comment (knocte) [reply tests/DotNetLightning.Core.Tests/TransactionTests.fs:352]: love that! I wonder if we could write an FSharpLint rule that could suggest this

File: tests/DotNetLightning.Core.Tests/TransactionTests.fs [hunk 6]
```diff
@@ -326,4 +329,551 @@ let testList = testList "transaction tests" [
...
+                remoteChannelPrivKeys
+                localCommitmentTx
+
+        let outputsSumFromRemoteSpendingLocalCommitment =
+            (Result.deref remoteSpendingLocalCommitmentRes.MainOutput)
+                .SendAll(remoteDestPubKey.WitHash.ScriptPubKey)
+                .SendFees(Money.Zero)
+                .BuildTransaction(true)
+                .Outputs
+                |> Seq.sumBy (fun output -> output.Value)
```
Comment (joemphilips) [inline tests/DotNetLightning.Core.Tests/TransactionTests.fs:579]: Must be better if you pick the one you intended instead of summing up?

```fsharp
let dest = remoteDestPubKey.WItHash.ScriptPubKey
// ...
    Seq.pick (fun output -> if output.ScriptPubKey = dest then Some (output.Value) else None) // instead of Seq.filter
```

File: tests/DotNetLightning.Core.Tests/TransactionTests.fs [hunk 7]
```diff
@@ -326,4 +329,551 @@ let testList = testList "transaction tests" [
...
+            
+        let expectedAmountFromToLocal =
+            let localAmount = lastLocalCommitmentSpec.ToLocal.ToMoney()
+            let fee = revokedLocalCommitmentTx.GetFee [| fundingScriptCoin :> ICoin |]
+            localAmount - fee
+        let expectedAmountFromToRemote =
+            lastLocalCommitmentSpec.ToRemote.ToMoney()
+
+
+        Expect.equal outputsSumFromRemoteSpendingRevokedRemoteCommitment (expectedAmountFromToRemote + expectedAmountFromToLocal) "wrong penalty tx"
```
Comment (joemphilips) [inline tests/DotNetLightning.Core.Tests/TransactionTests.fs:623]: You are running 3 attempt for `HandleFundingTxSpent`, and setup and assertions are mostly the same, it can be more succinct.

File: tests/DotNetLightning.Core.Tests/TransactionTests.fs (hunk 6 above)
Comment (aarani) [reply tests/DotNetLightning.Core.Tests/TransactionTests.fs]: Done.

File: tests/DotNetLightning.Core.Tests/TransactionTests.fs (hunk 5 above)
Comment (aarani) [reply tests/DotNetLightning.Core.Tests/TransactionTests.fs]: Done.

File: tests/DotNetLightning.Core.Tests/TransactionTests.fs (hunk 7 above)
Comment (aarani) [reply tests/DotNetLightning.Core.Tests/TransactionTests.fs]: I grouped validation, commitment tx creation (as much as I could) into functions

File: tests/DotNetLightning.Core.Tests/TransactionTests.fs [hunk 8]
```diff
@@ -326,4 +329,497 @@ let testList = testList "transaction tests" [
...
+            LocalCommit = remoteLocalCommit
+            RemoteCommit = remoteRemoteCommit 
+            LocalChanges = LocalChanges.Zero
+            RemoteChanges = RemoteChanges.Zero
+        }
+
+        let validateClosingResult (expectedAmount: Money) (errorMsg: string) (result: ClosingHelpers.ClosingResult) =
+            let dest =
+                let remoteDestPrivKey = new Key()
+                let remoteDestPubKey = remoteDestPrivKey.PubKey
```
Comment (joemphilips) [inline tests/DotNetLightning.Core.Tests/TransactionTests.fs:553]: nit: you don't have to create variable `remoteDestPubKey` and just use`remoteDestPrivKey.PubKey.WitHash.ScriptPubKey` directly in case like this.
The only case you should assign a new variable like this is to give a new information to the reader by its name.
e.g. 

```fsharp
// good example
let pubkeyToUseInMultisig = myPrivKey.PubKey // it is not self-evident from the right hand that this key is going to be used for multisig

// bad example
let pubkeyToUseInMultisig = myPrivKeyForMultisig.PubKey // no new information
```

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs [hunk 9]
```diff
@@ -17,3 +24,439 @@ module internal ChannelConstantHelpers =
...
+                toRemoteIndexOpt
+                |> Option.iter
+                    (
+                        fun toRemoteIndex ->
+                            let localPaymentPrivKey =
+                                perCommitmentPoint.DerivePaymentPrivKey
+                                    localChannelPrivKeys.PaymentBasepointSecret
+
+                            transactionBuilder.SetVersion TxVersionNumberOfCommitmentTxs
+                            |> ignore<TransactionBuilder>
```
Comment (joemphilips) [inline src/DotNetLightning.Core/Channel/ChannelHelpers.fs:303]: nit: why not just `ignore` ?

Comment (aarani) [reply src/DotNetLightning.Core/Channel/ChannelHelpers.fs:303]: We have a policy wrt this, the reason (short answer) I think is to not accidentally ignore a function (e.g. int -> TransactionBuilder) instead of the outcome/result of the function. @knocte can give you a more complete answer.
Ah also there's a tweet about this: https://twitter.com/ChetHusk/status/1351156012590526474

File: tests/DotNetLightning.Core.Tests/TransactionTests.fs (hunk 8 above)
Comment (aarani) [reply tests/DotNetLightning.Core.Tests/TransactionTests.fs]: Fixed

File: src/DotNetLightning.Core/Channel/ChannelHelpers.fs (hunk 9 above)
Comment (joemphilips) [reply src/DotNetLightning.Core/Channel/ChannelHelpers.fs:303]: Ah I see, I think I will follow this rule too from now on.

File: tests/DotNetLightning.Core.Tests/TransactionTests.fs (hunk 8 above)
Comment (joemphilips) [reply tests/DotNetLightning.Core.Tests/TransactionTests.fs]: > P.S: all these codes are taken from the same TransactionTests file so I didn't really pay more attention to this sort of stuff

I cannot believe I wrote code like this by myself ...

Comment (aarani) [reply tests/DotNetLightning.Core.Tests/TransactionTests.fs]: Yes, it was probably missed in previous reviews 👍🏻 
https://github.com/joemphilips/DotNetLightning/blob/master/tests/DotNetLightning.Core.Tests/TransactionTests.fs#L159

**Review by joemphilips — APPROVED** [review]
LGTM: a80670236314233253ef568e12705ce27f435402

**Comment by knocte** [comment]
> LGTM: https://github.com/joemphilips/DotNetLightning/commit/a80670236314233253ef568e12705ce27f435402

Rebasing from github UI (no changes).
